packages/hl7/src/client.ts

```diff
@@ -26,24 +29,79 @@ export class Hl7Client extends Hl7Base {
     this.port = this.options.port;
     this.encoding = this.options.encoding;
     this.keepAlive = this.options.keepAlive ?? false;
+    this.connectTimeout = this.options.connectTimeout ?? 30000; // Default 30 seconds
   }
 
   connect(): Promise<Hl7Connection> {
+    // If we already have a connection, use it
     if (this.connection) {
       return Promise.resolve(this.connection);
     }
 
+    // If there's an ongoing connection attempt, destroy it
+    if (this.socket) {
+      this.socket.removeAllListeners();
+      this.socket.destroy();
+      this.socket = undefined;
+    }
+
     return new Promise((resolve, reject) => {
-      const socket = connect({ host: this.host, port: this.port, keepAlive: this.keepAlive }, () => {
+      // Create the socket
+      this.socket = connect({
+        host: this.host,
+        port: this.port,
+        keepAlive: this.keepAlive,
+      });
+
+      // Set timeout if specified
+      if (this.connectTimeout > 0) {
+        this.socket.setTimeout(this.connectTimeout);
+
+        // Handle timeout event
+        this.socket.on('timeout', () => {
+          const error = new Error(`Connection timeout after ${this.connectTimeout}ms`);
+          if (this.socket) {
+            this.socket.destroy();
+            this.socket = undefined;
+          }
+          reject(error);
+        });
+      }
+
+      // Handle successful connection
+      this.socket.on('connect', () => {
+        if (!this.socket) {
+          return; // Socket was already destroyed
+        }
+
+        // Create the HL7 connection
         let connection: Hl7Connection;
-        this.connection = connection = new Hl7Connection(socket, this.encoding);
-        socket.off('error', reject);
-        connection.addEventListener('close', () => this.dispatchEvent(new Hl7CloseEvent()));
-        connection.addEventListener('error', (event) => this.dispatchEvent(new Hl7ErrorEvent(event.error)));
+        this.connection = connection = new Hl7Connection(this.socket, this.encoding);
+
+        // Remove the timeout listener as we're now connected
+        this.socket.setTimeout(0);
```

Comment:
Do we only wanna timeout whenever connecting?

It might be useful to be able to detect when a connection is broken, which by default won't happen, even when keepalive probes are not responded to for long periods of time 🤔

Reply:
https://nodejs.org/docs/latest-v22.x/api/net.html#socketsettimeouttimeout-callback

> Sets the socket to timeout after timeout milliseconds of inactivity on the socket. By default net.Socket do not have a timeout.

I don't think we want to timeout on inactivity though 🤔

Reply:
I mean if keepalive = true, then even if keepalive messages are not acked, we never do anything about fixing that broken connection. Typically keepalive is paired with a timeout in case of flaky connections. Otherwise you could be sending stuff into the void and have no knowledge of it until you send the next message

Reply:
I thought we set a timeout automatically when setting keepalive, otherwise it doesn't fix itself if the other side stops responding to keepalive messages

Reply:
Mm, i see, you mean the ping/pongs - sorry, I was misinterpreting "keepalives".

Maybe?  I don't have as much conviction on that point.

Reply:
Yeah I mean if we do that it can always be a separate PR, I think we mostly added keepalive to keep LB from timing out connections for the moment. So it serves that purpose as is